include/swift/AST/DeclContext.h

```diff
@@ -288,6 +288,38 @@ class alignas(1 << DeclContextAlignInBits) DeclContext
     return ParentAndKind.getInt() == ASTHierarchy::Decl ?
       reinterpret_cast<Decl*>(this + 1) : nullptr;
   }
+
+  /// Find the import that makes the given nominal declaration available.
+  Optional<AttributedImport<ImportedModule>> findImportFor(
+      ModuleDecl *nominalModule, ASTContext *nominalASTCtx) {
+    // If the nominal type is from the current module, there's no import.
+    if (nominalModule == getParentModule())
+      return None;
+
+    auto fromSourceFile = getParentSourceFile();
+    if (!fromSourceFile)
+      return None;
+
+    // Look to see if the owning module was directly imported.
+    for (const auto &import : fromSourceFile->getImports()) {
+      if (import.module.importedModule == nominalModule)
+        return import;
+    }
+
+    // Now look for transitive imports.
+    auto &importCache = nominalASTCtx.getImportCache();
+    for (const auto &import : fromSourceFile->getImports()) {
+      auto &importSet = importCache.getImportSet(import.module.importedModule);
+      for (const auto &transitive : importSet.getTransitiveImports()) {
+        if (transitive.importedModule == nominalModule) {
+          return import;
+        }
+      }
+    }
+
+    return None;
```

Comment:
Could you move the implementation into DeclContext.cpp? Generally unless the implementation is trivial, it's best not to put it in the header, as otherwise if you change it, any file that imports DeclContext.h will need to be re-compiled.

Reply:
Sure